### PR TITLE
(#3852) - New tests for replication events

### PR DESCRIPTION
**Pull request by daleharvey**

### CODE REVIEW

**Comment by daleharvey** [comment]
So these tests found one bug, it looks like the event listener fixes have fixed most of the problems with the multiple `paused` / `active`, I wouldnt be surprised if there were some more hidden cases but it looks like we are generally ok, will use these as the basis for testing https://github.com/pouchdb/pouchdb/issues/4251 too

**Comment by nolanlawson** [comment]
Awesome, +1.

File: tests/integration/test.replication_events.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            db.bulkDocs([{_id: 'e'}, {_id: 'f'}]);
+          } else if (counter === 6) {
+            // All good, wait for pause
+          } else {
+            done('active fired incorrectly');
+          }
+        });
+
+        repl.on('paused', function(err) {
+          counter++;
```
Comment (nolanlawson) [inline tests/integration/test.replication_events.js]: ^ apparently Sinon can get rid of this manual counter stuff.

Comment (daleharvey) [inline tests/integration/test.replication_events.js]: Its possible, I try to reserve sinon for actually mocking implemetation in unit tests, I find a lot of times when people put sinon in integration tests they hide fairly large errors, but will see how sinon is working in the other PR for the replication options

**Comment by daleharvey** [comment]
Sweet cheers - https://github.com/pouchdb/pouchdb/commit/a17be0693f325449ed7716d8dc49a7c3a0069b78

**Pull request closed without merging**